src/lib/statistics/cardinality_estimator.cpp

```diff
@@ -0,0 +1,858 @@
+#include "logical_query_plan/mock_node.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "logical_query_plan/sort_node.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+#include "logical_query_plan/union_node.hpp"
+#include "logical_query_plan/validate_node.hpp"
+#include "operators/operator_join_predicate.hpp"
+#include "operators/operator_scan_predicate.hpp"
+#include "resolve_type.hpp"
+#include "statistics/cardinality_estimation_cache.hpp"
+#include "statistics/column_statistics.hpp"
+#include "statistics/statistics_objects/equal_distinct_count_histogram.hpp"
+#include "statistics/statistics_objects/generic_histogram.hpp"
+#include "statistics/statistics_objects/generic_histogram_builder.hpp"
+#include "statistics/statistics_objects/single_bin_histogram.hpp"
+#include "storage/storage_manager.hpp"
+#include "storage/table.hpp"
+#include "table_statistics.hpp"
+#include "utils/assert.hpp"
+
+namespace {
+
+using namespace opossum;  // NOLINT
+
+template <typename T>
+std::optional<float> estimate_null_ratio_of_column(const TableStatistics& table_statistics,
+                                                   const ColumnStatistics<T>& column_statistics) {
+  // If the column has an explicit null value ratio associated with it, we can just use that
+  if (column_statistics.null_value_ratio) {
+    return column_statistics.null_value_ratio->ratio;
+  }
+
+  // Otherwise derive the NVR from the total count of an histogram (which excludes NULLs) and the TableStatistics
+  // row count (which includes NULLs)
+  if (column_statistics.histogram && table_statistics.row_count != 0.0f) {
+    return 1.0f - (static_cast<float>(column_statistics.histogram->total_count()) / table_statistics.row_count);
+  }
+
+  return std::nullopt;
+}
+
+}  // namespace
+
+namespace opossum {
+
+std::shared_ptr<AbstractCardinalityEstimator> CardinalityEstimator::new_instance() const {
+  return std::make_shared<CardinalityEstimator>();
+}
+
+Cardinality CardinalityEstimator::estimate_cardinality(const std::shared_ptr<AbstractLQPNode>& lqp) const {
+  const auto estimated_statistics = estimate_statistics(lqp);
+  return estimated_statistics->row_count;
+}
+
+std::shared_ptr<TableStatistics> CardinalityEstimator::estimate_statistics(
+    const std::shared_ptr<AbstractLQPNode>& lqp) const {
+  /**
+   * 1. Try a cache lookup for requested LQP
+   */
+  auto bitmask = std::optional<JoinGraphStatisticsCache::Bitmask>{};
+  if (cardinality_estimation_cache.join_graph_statistics_cache) {
+    bitmask = cardinality_estimation_cache.join_graph_statistics_cache->bitmask(lqp);
+    if (bitmask) {
+      const auto cached_statistics =
+          cardinality_estimation_cache.join_graph_statistics_cache->get(*bitmask, lqp->column_expressions());
+      if (cached_statistics) {
+        return cached_statistics;
+      }
+    }
+  }
+
+  if (cardinality_estimation_cache.statistics_by_lqp) {
```

Comment:
Added 
```c++

   * Lookup in `join_graph_statistics_cache` is expected to have a higher hit-rate (since every bitmask represents
   * multiple LQPs) than `statistics_by_lqp`. Thus lookup in `join_graph_statistics_cache` is performed first.
```

Reply:
“hit rate”

Reply:
yep